File: Go/Solutions/projeuler/Problem025.go
package euler

import (
	"fmt"
	"time"
)

/*
Problem025 answers the problem at : https://projecteuler.net/problem=25

* Problem 25:
The Fibonacci sequence is defined by the recurrence relation:

    Fn = Fn−1 + Fn−2, where F1 = 1 and F2 = 1.

Hence the first 12 terms will be:

    F1 = 1
    F2 = 1
    F3 = 2
    F4 = 3
    F5 = 5
    F6 = 8
    F7 = 13
    F8 = 21
    F9 = 34
    F10 = 55
    F11 = 89
    F12 = 144

The 12th term, F12, is the first term to contain three digits.

What is the index of the first term in the Fibonacci sequence to contain 1000 digits?
*/
func Problem025() {
	start := time.Now()
	ans := 0
	for true {
		if numLen(Fibonacci(int64(ans))) == 1000 {
			break
		}
		ans++
	}
	end := time.Now()

	fmt.Printf("\nAnswer to Problem 25 : %d\n", ans)
	fmt.Printf("Time Taken: %f seconds\n\n", end.Sub(start).Seconds())
}

File: Python/Problem025.py
from functools import lru_cache
from time import time

''' 
Problem 25: https://projecteuler.net/problem=25



The F(ibonacci) sequence is def(ined) by the recurrence relation:

        F(n) = F(n−1) + F(n−2), where F(1) = 1 and F(2) = 1.

Hence the f(irst) 12 terms will be:

        F(1) = 1
        F(2) = 1
        F(3) = 2
        F(4) = 3
        F(5) = 5
        F(6) = 8
        F(7) = 13
        F(8) = 21
        F(9) = 34
        F(10) = 55
        F(11) = 89
        F(12) = 144

The 12th term, F(12), is the f(irst) term to contain three digits.

What is the index of( the) f(irst) term in the F(ibonacci) sequence to contain 1000 digits?
'''

@lru_cache
def fibonacci(n):
    if n in [0, 1]:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)

i, start = 0, time()
while len(str(fibonacci(i))) < 1000:
    i+=1

if( __name__) == '__main__':
    print(f"\nAnswer: { i }")
    print(f"Time Taken: { time() - start }\n")